### PR TITLE
[HttpFoundation] Do not swallow trailing `=` in cookie value

### PR DESCRIPTION
**Pull request by OskarStark**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix https://github.com/symfony/symfony/issues/51814
| License       | MIT

cc @pschultz as you opened the bug report

### CODE REVIEW

File: src/Symfony/Component/HttpFoundation/Tests/CookieTest.php [hunk 1]
```diff
@@ -324,6 +324,9 @@ public function testFromString()
         $cookie = Cookie::fromString('foo=bar', true);
         $this->assertEquals(Cookie::create('foo', 'bar', 0, '/', null, false, false, false, null), $cookie);
 
+        $cookie = Cookie::fromString('foo=bar=', true);
```
Comment (OskarStark) [inline src/Symfony/Component/HttpFoundation/Tests/CookieTest.php:327]: This test case was failing before

**Comment by OskarStark** [comment]
Or should this be fixed inside HeaderUtils::split() method?

**Comment by stof** [comment]
I guess we have similar bugs in other places using `HeaderUtils::split` with multiple separator (i.e. all of them). That's because HeaderUtils considers that all separators are list separators. But `=` is a key-value separator.

**Comment by OskarStark** [comment]
So do you have a proposal on where to fix this?

**Comment by stof** [comment]
Well, HeaderUtils should be the place to fix it, but not sure we can fix it with the existing signature (unless we hardcode a special behavior for `=`).

**Comment by nicolas-grekas** [comment]
The bug is clearly in HeaderUtils, see
https://github.com/symfony/symfony/blob/adcd3d06d078c86618ec3b27892e474483c63895/src/Symfony/Component/HttpFoundation/Tests/HeaderUtilsTest.php#L48

`foo_cookie=foo==` is correctly parsed while `foo_cookie=foo=` isn't

**Comment by OskarStark** [comment]
Will check on Monday 👍🏻

**Comment by OskarStark** [comment]
Review please

File: src/Symfony/Component/HttpFoundation/HeaderUtils.php [hunk 2]
```diff
@@ -257,10 +257,20 @@ private static function groupParts(array $matches, string $separators, bool $fir
         $i = 0;
         $partMatches = [];
         $previousMatchWasSeparator = false;
-        foreach ($matches as $match) {
+        foreach ($matches as $key => $match) {
             if (!$first && $previousMatchWasSeparator && isset($match['separator']) && $match['separator'] === $separator) {
                 $previousMatchWasSeparator = true;
                 $partMatches[$i][] = $match;
+            } elseif (!$first && !isset($match['separator']) && isset($matches[$key + 1], $matches[$key + 1]['separator']) && $matches[$key+1]['separator'] === $separator && !isset($matches[$key+2])) { // ends with a separator, so it must be part of the value
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/HeaderUtils.php]:
```suggestion
            } elseif (!$first && !isset($match['separator']) && ($matches[$key + 1]['separator'] ?? null) === $separator && !isset($matches[$key + 2])) { // ends with a separator, so it must be part of the value
```

File: src/Symfony/Component/HttpFoundation/HeaderUtils.php [hunk 3]
```diff
@@ -257,10 +257,20 @@ private static function groupParts(array $matches, string $separators, bool $fir
...
+                $numberOfSeparators = 0;
+                for ($i = $key+1; $i < \count($matches); ++$i) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/HeaderUtils.php]: do we need this loop? didn't we rule out with `!isset($matches[$key+2])`?

**Review by nicolas-grekas — COMMENTED** [review]
The whole logic is opaque to me :sweat:

File: src/Symfony/Component/HttpFoundation/HeaderUtils.php (hunk 2 above)
Comment (OskarStark) [reply src/Symfony/Component/HttpFoundation/HeaderUtils.php:264]: Yes

File: src/Symfony/Component/HttpFoundation/HeaderUtils.php (hunk 3 above)
Comment (OskarStark) [reply src/Symfony/Component/HttpFoundation/HeaderUtils.php]: Yes we rule out, but we don't know if one or 2 trailing `=`

I agree, its very opaque and I really don't like the whole code :-(

File: src/Symfony/Component/HttpFoundation/HeaderUtils.php (hunk 2 above)
Comment (OskarStark) [reply src/Symfony/Component/HttpFoundation/HeaderUtils.php]: Works

**Review by nicolas-grekas — APPROVED** [review]
I rewrote the implementation, it now makes sense, at least to me :)
AND, it handles more edge cases.

**Comment by OskarStark** [comment]
Thank you Nicolas!! 👏

File: src/Symfony/Component/HttpFoundation/Tests/HeaderUtilsTest.php [hunk 4]
```diff
@@ -46,13 +46,15 @@ public static function provideHeaderToSplit(): array
 
             [[['foo_cookie', 'foo=1&bar=2&baz=3'], ['expires', 'Tue, 22-Sep-2020 06:27:09 GMT'], ['path', '/']], 'foo_cookie=foo=1&bar=2&baz=3; expires=Tue, 22-Sep-2020 06:27:09 GMT; path=/', ';='],
             [[['foo_cookie', 'foo=='], ['expires', 'Tue, 22-Sep-2020 06:27:09 GMT'], ['path', '/']], 'foo_cookie=foo==; expires=Tue, 22-Sep-2020 06:27:09 GMT; path=/', ';='],
```
Comment (OskarStark) [inline src/Symfony/Component/HttpFoundation/Tests/HeaderUtilsTest.php:48]:
```suggestion
            [[['foo_cookie', 'foo==='], ['expires', 'Tue, 22-Sep-2020 06:27:09 GMT'], ['path', '/']], 'foo_cookie=foo===; expires=Tue, 22-Sep-2020 06:27:09 GMT; path=/', ';='],
            [[['foo_cookie', 'foo=='], ['expires', 'Tue, 22-Sep-2020 06:27:09 GMT'], ['path', '/']], 'foo_cookie=foo==; expires=Tue, 22-Sep-2020 06:27:09 GMT; path=/', ';='],
```

Does this pass as well?

Comment (nicolas-grekas) [reply src/Symfony/Component/HttpFoundation/Tests/HeaderUtilsTest.php:48]: yes

**Comment by nicolas-grekas** [comment]
Thank you @OskarStark.

**Comment by OskarStark** [comment]
Thank you, as you did the reimplementation 👍
